src/execution/operator/join/physical_piecewise_merge_join.cpp

```diff
@@ -144,43 +268,75 @@ SinkFinalizeType PhysicalPiecewiseMergeJoin::Finalize(Pipeline &pipeline, Event
 //===--------------------------------------------------------------------===//
 class PiecewiseMergeJoinState : public OperatorState {
 public:
-	explicit PiecewiseMergeJoinState(const PhysicalPiecewiseMergeJoin &op)
-	    : op(op), first_fetch(true), finished(true), left_position(0), right_position(0), right_chunk_index(0) {
+	explicit PiecewiseMergeJoinState(const PhysicalPiecewiseMergeJoin &op, BufferManager &buffer_manager,
+	                                 bool force_external)
+	    : op(op), buffer_manager(buffer_manager), force_external(force_external), left_position(0), first_fetch(true),
+	      finished(true), right_position(0), right_chunk_index(0) {
 		vector<LogicalType> condition_types;
 		for (auto &cond : op.conditions) {
 			lhs_executor.AddExpression(*cond.left);
 			condition_types.push_back(cond.left->return_type);
 		}
-		join_keys.Initialize(condition_types);
+		lhs_keys.Initialize(condition_types);
 		if (IsLeftOuterJoin(op.join_type)) {
-			left_found_match = unique_ptr<bool[]>(new bool[STANDARD_VECTOR_SIZE]);
-			memset(left_found_match.get(), 0, sizeof(bool) * STANDARD_VECTOR_SIZE);
+			lhs_found_match = unique_ptr<bool[]>(new bool[STANDARD_VECTOR_SIZE]);
+			memset(lhs_found_match.get(), 0, sizeof(bool) * STANDARD_VECTOR_SIZE);
 		}
+		lhs_layout.Initialize(op.children[0]->types);
 	}
 
 	const PhysicalPiecewiseMergeJoin &op;
+	BufferManager &buffer_manager;
+	bool force_external;
+
+	// Block sorting
+	DataChunk lhs_keys;
+	ExpressionExecutor lhs_executor;
+	unique_ptr<bool[]> lhs_found_match;
+	RowLayout lhs_layout;
+	unique_ptr<LocalSortState> lhs_local_state;
+	unique_ptr<GlobalSortState> lhs_global_state;
+	idx_t lhs_count;
+	idx_t lhs_has_null;
+
+	// Simple scans
+	idx_t left_position;
+
+	// Complex scans
 	bool first_fetch;
 	bool finished;
-	idx_t left_position;
 	idx_t right_position;
 	idx_t right_chunk_index;
-	DataChunk left_chunk;
-	DataChunk join_keys;
-	MergeOrder left_orders;
-	//! The executor of the RHS condition
-	ExpressionExecutor lhs_executor;
-	unique_ptr<bool[]> left_found_match;
+	idx_t right_base;
 
 public:
 	void ResolveJoinKeys(DataChunk &input) {
 		// resolve the join keys for the input
-		join_keys.Reset();
-		lhs_executor.SetChunk(input);
-		join_keys.SetCardinality(input);
-		for (idx_t k = 0; k < op.conditions.size(); k++) {
-			lhs_executor.ExecuteExpression(k, join_keys.data[k]);
-			// sort by join key
-			OrderVector(join_keys.data[k], join_keys.size(), left_orders);
+		lhs_keys.Reset();
+		lhs_executor.Execute(input, lhs_keys);
+
+		// Count the NULLs so we can exclude them later
+		// TODO: Sort any multi-comparison NULLs to the end using an initial sort column
+		lhs_count = lhs_keys.size();
+		for (auto &key : lhs_keys.data) {
+			lhs_has_null = lhs_count - key.CountValid(lhs_count);
```

Comment:
Am I missing something, or could this be replaced with `VectorOperations::HasNull`?